### DISCUSSION TITLE
Issues using system piping for error logging

### DISCUSSION THREAD

**Discussion by fallible-algebra** [bevyengine/bevy#8637] [Q&A]
I've been working on a bevy project for a couple of months and have come across a minor information hiccup with error logging. 

When using system piping for error logging, the results the system return will get reported by the logging suite as coming from the wrapper system (i.e. [`warn`](https://docs.rs/bevy/latest/bevy/prelude/fn.warn.html) or [`info`](https://docs.rs/bevy/latest/bevy/prelude/fn.info.html)) rather than the function that returned the error i.e. as a stripped down version of [this example](https://github.com/bevyengine/bevy/blob/25f013ba1b809e9f962dafbefaeb860512f386e8/examples/ecs/system_piping.rs)

```rust
fn main() {
    App::new()
    // ...
    .add_system(result_system.pipe(warn)
    .run();
}

pub mod where_result_system_is {
    pub fn result_system(/* whatever inputs */) -> Result<(), String> {
        Err("Error returned by system".to_owned())
    }
}
```

```shell
# Actual output, shows the path to where `warn` is defined.
WARN bevy_ecs::system::system_piping::adapter: Error returned by system
# Desired output, shows the path to where `result_system` is defined.
WARN my_project::where_result_system_is: Error returned by system
```

The latter is desired as it gives me information as to where the warning is actually coming from, and my project has a wide and deep directory tree.

I'm aware this is behaviour to be expected from using the logging suite + the system piping interfaces together, but this is suboptimal from a point of view of using system piping for error logging. Is there a macro-based approach someone else has worked on? Something that transforms a system that returns a result into a system that doesn't return a result that matches on the result in-function-body, i.e.

```rust
#[system_result_transform_macro_of_some_kind(warn)]
pub fn result_system(/* whatever inputs */) -> Result<(), String> {
    Err("Error returned by system")
}
// Becomes
pub fn result_system(/* whatever inputs */) {
    // Or however else the body needs to be encapsulated to allow i.e. ? syntax.
    let res = {Err("Error returned by system")};
    match res {
        Ok(()) => {},
        Err(err) => warn!(err)
    }
}
// So that error location is preserved.
```

Or is there something else I should be relying on, have I missed something obvious, or is this all splitting hairs? Thanks.

**Comment by alice-i-cecile** [comment]
I agree that this would be a much more useful result. Can you reproduce it on `main` and then open an issue?

**Reply by fallible-algebra** [reply]
Have done, thank you! https://github.com/bevyengine/bevy/issues/8638
